File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/domain/user/Email.java
package br.com.deltaglobalbank.identity.domain.user;

import java.util.regex.Pattern;

public record Email(String value) {

    public static final int MAX_LENGTH = 255;
    private static final Pattern EMAIL_REGEX =
        Pattern.compile("^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$");

    public Email {
        String normalizedEmail = value.trim().toLowerCase();
        if (normalizedEmail.isBlank()) {
            throw new IllegalArgumentException("email_blank");
        }
        if (normalizedEmail.length() > MAX_LENGTH) {
            throw new IllegalArgumentException("email_too_long");
        }
        if (!EMAIL_REGEX.matcher(normalizedEmail).matches()) {
            throw new IllegalArgumentException("email_invalid");
        }
        value = normalizedEmail;
    }

    @Override
    public String toString() {
        return value;
    }
}

File: services/customers-service/src/main/kotlin/br/com/deltaglobalbank/customers/domain/customer/valueobjects/Email.kt
package br.com.deltaglobalbank.customers.domain.customer.valueobjects

@JvmInline
value class Email private constructor(val value: String) {

    companion object {
        operator fun invoke(rawEmail: String): Email {
            val normalizedEmail = rawEmail.trim().lowercase()
            require(normalizedEmail.isNotBlank()) { "email_blank" }
            require(normalizedEmail.length <= MAX_LENGTH) { "email_too_long" }
            require(EMAIL_REGEX.matches(normalizedEmail)) { "email_invalid" }
            return Email(normalizedEmail)
        }

        const val MAX_LENGTH = 255
        private val EMAIL_REGEX = Regex("^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$")
    }

    override fun toString(): String = value
}